rasa/core/policies/rule_policy.py

```diff
@@ -896,33 +1015,38 @@ def predict_action_probabilities(
         # Rasa Open Source default actions overrule anything. If users want to achieve
         # the same, they need to write a rule or make sure that their loop rejects
         # accordingly.
-        default_action_name = self._find_action_from_default_actions(tracker)
+        (
+            default_action_name,
+            self._prediction_source,
+        ) = self._find_action_from_default_actions(tracker)
 
         # text has priority over intents including default,
         # however loop happy path has priority over rules prediction
         if default_action_name and not rules_action_name_from_text:
-            self._prediction_source = DEFAULT_RULES
-            return self._prediction(
+            return self._rule_prediction(
                 self._prediction_result(default_action_name, tracker, domain)
             )
 
         # A loop has priority over any other rule except defaults.
         # The rules or any other prediction will be applied only if a loop was rejected.
         # If we are in a loop, and the loop didn't run previously or rejected, we can
         # simply force predict the loop.
-        loop_happy_path_action_name = self._find_action_from_loop_happy_path(tracker)
+        (
+            loop_happy_path_action_name,
+            self._prediction_source,
+        ) = self._find_action_from_loop_happy_path(tracker)
         if loop_happy_path_action_name:
-            self._prediction_source = LOOP_RULES
             # this prediction doesn't use user input
             # and happy user input anyhow should be ignored during featurization
-            return self._prediction(
+            return self._rule_prediction(
                 self._prediction_result(loop_happy_path_action_name, tracker, domain),
                 is_no_user_prediction=True,
             )
 
         # predict rules from text first
         if rules_action_name_from_text:
-            return self._prediction_with_unhappy_path(
+            self._prediction_source = prediction_source_from_text
```

Comment:
We can't set any policy state during prediction time. This would e.g. blow up if a second user message is handled in parallel by the policy.

Reply:
It is like this in `master`, but I don't think is actually used during prediction time

Reply:
great point, totally slipped out of my mind. Thank you for spotting, it saved us a lot of trouble!